### PR TITLE
Slonigiraf grant application

### PR DESCRIPTION
**Pull request by slonigiraf**
### Project Abstract

Slonigiraf is an p2p education system.
We want to bring to Web3 infrastructure the renowned Slonigiraf learning game, which has been designed to deliver more interactive and productive lessons through peer-to-peer learning and quality control through cross-exams.

### For which grant [level](https://github.com/w3f/Grants-Program#level_slider-levels) are you applying? 
- [x] **Level 1**:  Up to $10,000, 2 approvals
- [ ] **Level 2**:  Up to $50,000, 3 approvals
- [ ] **Level 3**:  Unlimited, 5 approvals (for > $100k Web3 Foundation Council approval)

### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied, renamed ( `project_name.md`) and updated.
- [x] A BTC or Ethereum (DAI/USDT) address for the payment of the milestones is provided inside the application.
- [x] I have read and acknowledged the [terms and conditions](https://github.com/w3f/Grants-Program/blob/master/docs/T&Cs.md).
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process).

### CODE REVIEW

**Comment by semuelle** [comment]
Thank you for your application, @slonigiraf. Some questions: 

- Have you considered implementing this as smart contracts? I currently don't see the need for a parachain around it. 
- Has this concept ever been done, off-chain? Has this idea been studied? I understand the usefulness of incentivising good teachers, but on the other hand this threatens teachers with insecure income.
- You state that " All parties are economically motivated to behave honestly", but what is the economic incentive of anyone _not_ to penalize a teacher?
- What is the purpose of your own token, other than donations?

**Comment by slonigiraf** [comment]
@semuelle thanks for questions. Here are the answers:
1. Have you considered implementing this as smart contracts? I currently don't see the need for a parachain around it.
> Yes, we initially thought that smart contracts on Ethereum/Polka etc would be enough. But there are two problems with such approach: UX and technical. 
UX one requires users to buy the token that allows to pay for smart contract execution. Most of our users don't know anything about crypto, thus it's exceptionally hard to make them buy some crypto. SLON token initially is given for free to any school/university.
Technical. We implement the system that should process requests from millions of users every day. No single chain can process transactions if all pupils in Russia will use our educational model at classes - too many transactions, even considering that we use offchain as much as possible. Thus we look for any approach to shard and parallelize transaction processing. Polkadot design with relay and parachains looks great to solve this parallelization issue.

2. Has this concept ever been done, off-chain? Has this idea been studied? 
> Yes the idea initially was developed in 2019 year at a private school as a paper game, see
https://slon-i-giraf.ru/app/work?view=paperGameView&language=ENG
Then in 2020 we implemented the web version of this educational model on Java/Mysql and use it for now.
Some of our classes are online and free to participate and you can join and see how does it work - every Saturday we have free to join biology class where students teach and examine each other using of Java/Mysql version of system. Here is our telegram group of the class: https://t.me/bio_slonigiraf and here is a Zoom link (20.00 each Saturday, Moscow time): https://t.me/bio_slonigiraf

3. I understand the usefulness of incentivizing good teachers, but on the other hand this threatens teachers with insecure income.
> In different professions two models of payments exist: per hour (or per month) and per result. The first one is secure to workers but doesn't motivate them to produce the result. The second one - paying for the result enables to democratize the service market - you just pay for what you need and don't require CVs of worker. Also it enables us to include students and pupils in system as teachers - students teach each other and pay each other only for the result.

4. You state that " All parties are economically motivated to behave honestly", but what is the economic incentive of anyone not to penalize a teacher?
> The person who penalize teacher gets SLON from such teacher.

5. You state that "SLON coins play the role of a teacher's digital reputation", but also "You can, for example, put 100 SLON on your account". This sounds like I can buy reputation. 
> Yes, you can buy reputation - this is a system main idea. Since you pay for your reputation, you value it.

6. What is the relation between SLONs and guarantees? I can issue 5 guarantees for 100 SLON, but each guarantee, if penalized, costs me 100 SLON?
> Yes, you got it right

**Comment by slonigiraf** [comment]
I added a commit.

**Comment by semuelle** [comment]
Thanks for the update, @slonigiraf.  

I have read your comments and updates. If I understand correctly, there are two possible transactions per course, issuing insurance and buying it (punishing the teacher). Even if every student and every course uses the chain, I don't see how that could amount to "a huge number of transactions per second". I also disagree with the concept of buying reputation, because it will put those who can afford losing it at an advantage. For these reasons, I will not support your application. However, I am going to share it with the rest of my colleagues, and they are free to ask further questions and give their opinion.

**Comment by Lederstrumpf** [comment]
> UX one requires users to buy the token that allows to pay for smart contract execution. Most of our users don't know anything about crypto, thus it's exceptionally hard to make them buy some crypto. SLON token initially is given for free to any school/university.

You can also give out tokens for smart contract execution for free, if you're already on that course. Against the cost of you securing a parachain (very high if you just hand out tokens for free), this is much cheaper.

> Has this concept ever been done, off-chain? Has this idea been studied?
> ...

Can you link to studies that support the efficacy of this concept?

File: applications/slonigiraf.md [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+  Sarah then goes to Ed (the employer) to get a job that requires skill_1 and skill_2. Sarah presents Ed with
+  certificates and give Ed a chance to punish Tanya and Terry. If Ed believes Sarah has the skills, he will hire Sarah
+  and also pay Sarah for the certificates use. Otherwise, Ed may penalize teachers for incorrect certificates.\
+  SLON works as a reputation token. You could be surprised that we provide a possibility to buy reputation but this design
+  really works - since you pay for your reputation, you value it.\
+  The relation between SLONs and guarantees is following: teacher can issue 5 guarantees having 100 SLONs on his account, 
+  but each guarantee, if penalized, costs teacher 100 SLON.\
+  All parties: Sarah, Tanya, Teri and Ed are economically motivated to behave honestly.\
+  Ed is economic incentivized penalize a teacher in case if Sarah lack the skill because he gets SLON from such teacher.\
+  But also Ed can't behave dishonestly. If he starts to penalize teachers in case when skill is present - he will lose
```
Comment (Lederstrumpf) [inline applications/slonigiraf.md:30]: The reasoning here is not sound, and your application also does not describe how Ed attains an identity on the chain that entitles him to slash teachers. More generally, I struggle to see anything in this application that actually enforces the desired behaviours, not to mention prevention of bribery.

File: applications/slonigiraf.md [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+  all new workers - nobody whants to work with unfair hirer.\
+  In different professions two models of payments exist: per hour (or per month) and per result. The first one is secure to workers but doesn't motivate them to produce the result. The second one - paying for the result enables to democratize the service market - you just pay for what you need and don't require CVs of worker. Also it enables us to include students and pupils in system as teachers - students teach each other and pay each other only for the result.\
```
Comment (Lederstrumpf) [inline applications/slonigiraf.md:32]: I support what you're aiming to achieve here, but feel that a concrete design of incentivizing the correct behavior / punish misbehavior is missing.

As mentioned earlier, a parachain seems like a bad fit for this project, given the cost in securing one.

As such, I do not support the application in its current state.

**Review by Lederstrumpf — CHANGES_REQUESTED** [review]

**Comment by slonigiraf** [comment]
> Thanks for the update, @slonigiraf.
> 
> I have read your comments and updates. If I understand correctly, there are two possible transactions per course, issuing insurance and buying it (punishing the teacher). Even if every student and every course uses the chain, I don't see how that could amount to "a huge number of transactions per second". 

Let's calculate the required number of transactions per second.
There are two types of transactions that are sent: first - usual for any blockchain - transfer of SLON between holders. We will not count this type of transaction because the number of second type transactions is large enough to convince that we need a parallelization.
The second type transaction is a reimbursement where the person who is eligible to punish the teacher uses that right.
We have enough statistics from the school that uses such educational approach to estimate number of reimbursement transactions per second. Usually we process about 30 reimbursements per 45 minutes lesson per class of 20 students or 0.67 transactions per minute per 20 students.
In Russia there are about 15 million of pupils at schools. If we assume that all teachers will use such system at lesson we get (0.67*15*10^6/20) transactions per minute or 8375 transactions per second.
Substrate gives about 1000 transactions per second per chain. Thus we need at least 9 chains only for Russia to make this blockchain work. One can argue that in Russia there are different time zones and transaction load will be lowered by this mean. But don't forget that on Earth we have about 50x more pupils that in Russia.

**Comment by slonigiraf** [comment]
> efficacy of this concept
There is an ongoing study to compare the efficacy of such learning to traditional lecture based approach. We're looking for the comparison results to be published by the end of 2022 year.\
By the way working in pairs is known to improve the learging in schools, see for example paper on programming teaching (https://www.sciencedirect.com/science/article/abs/pii/S0360131520302219?dgcid=rss_sd_all). Thus it's highly likely that adding the quality controls for such work in pairs is a further improvement.

**Comment by slonigiraf** [comment]
> I also disagree with the concept of buying reputation, because it will put those who can afford losing it at an advantage.

Sebastian, this is how current education system is already organized. Universities issue diploma and risk to lose their reputation by graduating students with bad skills. We propose just accelerate that feedback loop for those who issue diploma and lower a barrier to enter market of diploma issuing to make it more democratic and competitive.

File: applications/slonigiraf.md (hunk 1 above)
Comment (slonigiraf) [reply applications/slonigiraf.md]: Let's consider different ways of misbehavior of parties.
We should note here that Sara and Tanya, Sara and Teri are persons who trust each other, and Ed is a public company that Sara trusts. This "trust" is an essential composite of the system. Teacher will not issue a certificate for a student whom he doesn't trust. A student will not show a certificate to a company that is known for misbehavior.
A). Sarah gets a certificate from Tanya but then intentionally tells Teri that she lack such skill. Teri penalize Tanya and Tanya loses some amount of SLON. As Tanya knows exactly that Sarah is good at the skill then Tanya decides not to work any more with Sarah and not to issue other certificates for Sarah. Sarah loses faith in herself in Tanya's mind. And this "faith" cost a lot. By the way Sarah has no motif to act in such way.
B). The same as above but Sarah allows Ed to penalize Tanya. Sarah will not get a work if she do so. It's not profitable for Sarah.
C). Ed knows that Sarah has a skill but penalize a teacher. He will lose all new workers - nobody wants to work with unfair hirer D). Teri knows that Sarah has a skill but penalize Tanya. Sarah knows that Teri cheated and will not work with him in future, Teri will lose a profit from issuing future certificates for Sarah.
E). Sarah bribes a teacher to issue a certificate. Ed will examine her and penalize the teacher. The pennalization sum is larger that Sarah's bribe. Thus the teacher can't afford to get a bribe from a student.

**Comment by slonigiraf** [comment]
Dear colleagues,  could you review the changes?

**Review by Noc2 — CHANGES_REQUESTED** [review]
I see two major issues with this application, which is why I haven’t approved it. Sorry for not sharing my feedback here earlier: how is the identity handled on-chain? Otherwise the system can easily be sybil attacked? @Lederstrumpf already mentioned this above. Apart from that I don’t think the concept that different teachers can effectively slash each other makes sense. Teri can in my opinion easily slash Tanya and no one will notice this, because it’s almost impossible to prove that teri actually did this on purpose. Knowledge, skills isn’t something you can easily measure in most cases.  It might also depend on the situation (for example someone is currently nervous), etc..  Furthermore you often have specific teachers for specific things, so the concept of they won’t do xyz, because this way you don’t ask the teacher to help you in the future doesn’t work for alot of cases.

File: applications/slonigiraf.md [hunk 3]
```diff
@@ -27,14 +27,20 @@
...
 every Saturday we have free to join biology class where students teach and examine each other 
-using of Java/Mysql version of system. [Here](https://t.me/bio_slonigiraf) is our telegram group of the class. And [here](https://us02web.zoom.us/j/86918554814?pwd=aUpndVI0dU9KclNybk1pVlM3K0FkQT09) is a Zoom link (20.00 each Saturday, Moscow time).
+using of Java/Mysql version of system. [Here](https://t.me/bio_slonigiraf) is our telegram group of the class. And [here](https://us02web.zoom.us/j/86918554814?pwd=aUpndVI0dU9KclNybk1pVlM3K0FkQT09) is a Zoom link (20.00 each Saturday, Moscow time). There is an ongoing study to compare the efficacy of such learning to traditional lecture based approach. We're looking for the comparison results to be published by the end of 2022 year.\
+By the way working in pairs is known to improve the learging in schools, see for example paper on [programming teaching](https://www.sciencedirect.com/science/article/abs/pii/S0360131520302219?dgcid=rss_sd_all). Thus it's highly likely that adding the quality controls for such work in pairs is a further improvement.
```
Comment (Lederstrumpf) [inline applications/slonigiraf.md:43]: This is unrelated.

File: applications/slonigiraf.md [hunk 4]
```diff
@@ -50,8 +50,14 @@ using of Java/Mysql version of system. [Here](https://t.me/bio_slonigiraf) is ou
...
-
+Technical. We implement the system that should process requests from millions of users every day.\
+Let's calculate the required number of transactions per second.\
+There are two types of transactions that are sent: first - usual for any blockchain - transfer of SLON between holders. We will not count this type of transaction because the number of second type transactions is large enough to convince that we need a parallelization.\
+The second type transaction is a reimbursement where the person who is eligible to punish the teacher uses that right.\
+We have enough statistics from the school that uses such educational approach to estimate number of reimbursement transactions per second. Usually we process about 30 reimbursements per 45 minutes lesson per class of 20 students or 0.67 transactions per minute per 20 students.\
```
Comment (Lederstrumpf) [inline applications/slonigiraf.md:63]: Am I understanding correctly that for a running class of 20 students, you on average have 30 reimbursements per 45 minutes lesson? That is, on average, every student requests 1.5 reimbursements per lesson? What are you putting these kids through? ;)

File: applications/slonigiraf.md [hunk 5]
```diff
@@ -50,8 +50,14 @@ using of Java/Mysql version of system. [Here](https://t.me/bio_slonigiraf) is ou
...
+In Russia there are about 15 million of pupils at schools. If we assume that all teachers will use such system at lesson we get (0.67*15*10^6/20) transactions per minute or 8375 transactions per second.\
+Substrate gives about 1000 transactions per second per chain. Thus we need at least 9 chains only for Russia to make this blockchain work. One can argue that in Russia there are different time zones and transaction load will be lowered by this mean. But don't forget that on Earth we have about 50x more pupils than in Russia.\
```
Comment (Lederstrumpf) [inline applications/slonigiraf.md:65]: If you need this high a tps just for reimbursements, then your issue is the reimbursement model with every person requesting on average 1.5 reimbursements per lesson. It is extremely unlikely you would secure 9 parachain slots for this, so I'd urge you to redesign your approach.

File: applications/slonigiraf.md [hunk 6]
```diff
@@ -50,8 +50,14 @@ using of Java/Mysql version of system. [Here](https://t.me/bio_slonigiraf) is ou
...
+Thus no single chain can process transactions if all pupils in Russia will use our educational model at classes - too many transactions, even considering that we use offchain as much as possible. Thus we look for any approach to shard and parallelize transaction processing. Polkadot design with relay and parachains looks great to solve this parallelization issue.
```
Comment (Lederstrumpf) [inline applications/slonigiraf.md:66]: not if you need 9 parachains only for Russia.

**Review by Lederstrumpf — CHANGES_REQUESTED** [review]
Hi @slonigiraf: My last review's comments were not addressed, barring wiping the issue of correct incentive design under the rug of "all involved parties trust one another" - if they do, there's little need for a blockchain.
And if you require at least 9 parachains to cover only Russia for reimbursements, then I fear my remark that "a parachain seems like a bad fit for this project, given the cost in securing one." has even greater weight.
Please also see my other comments in this review.

As such, I still do not support this application.

**Comment by slonigiraf** [comment]
> I see two major issues with this application, which is why I haven’t approved it. Sorry for not sharing my feedback here earlier: how is the identity handled on-chain? Without a proper identity system the rest of the solution can easily be sybil attacked. @Lederstrumpf already mentioned this above. Apart from that I don’t think the concept that different teachers can effectively slash each other makes sense. Teri can in my opinion easily slash Tanya and no one will notice this, because it’s almost impossible to prove that teri actually did this on purpose. 

The identity of users is tracked as follows: each certificate contains the address of the teacher and student. There is a balance known to all on the teacher's address. If the teacher's balance becomes less than the amount of the fine for the certificate, this certificate is considered temporarily canceled. If a student wants to show an active certificate to an employer to get a job, or to another teacher in order to take a second exam, the student signs that he agrees to grant that person the right to impose fines. The student knows the real identity and addresses of employee and teachers, teachers know a real identity and blockchain address of the student. Teacher knows the address who fined him and a real identity of the students who had the certificate. This information is enough to prevent "silent" fines. Teacher can track each fine with aid from the student. Sybil attack is not possible in principle in such system, as all transactions are done between real world identities.

**Comment by slonigiraf** [comment]
> 



> Knowledge, skills isn’t something you can easily measure in most cases. It might also depend on the situation (for example someone is currently nervous), etc..

A "skill" discussed here is a specific learned ability to perform an action that can can be attested by different examiners in a similar way (not all everyday skills fall in this definition).

**Comment by slonigiraf** [comment]
> Furthermore you often have specific teachers for specific things, so the concept of they won’t do xyz, because this way you don’t ask the teacher to help you in the future doesn’t work for alot of cases.

For sure our educational system doesn't cover all people educational interaction types. It's just an instrument that was shown to successfully work for example in schools. For sure in some cases like you described - it will not work. But this doesn't mean that we don't need the instrument by itself. We just need to use the instrument to those applications where it's useful.

**Comment by slonigiraf** [comment]
> And if you require at least 9 parachains to cover only Russia for reimbursements, then I fear my remark that "a parachain seems like a bad fit for this project, given the cost in securing one." has even greater weight.
> Please also see my other comments in this review.

As we need a lot of parachains to make the system work it's unlikely to include all of them in Polkadot or Kusama directly instead we plan to use our own relay chain with our parachains. Our relay chain can be connected to Polkadot or Kusama with bridges.

File: applications/slonigiraf.md (hunk 5 above)
Comment (slonigiraf) [reply applications/slonigiraf.md:67]: As we need a lot of parachains to make the system work it's unlikely to include all of them in Polkadot or Kusama directly instead we plan to use our own relay chain with our parachains. Our relay chain can be connected to Polkadot or Kusama with bridges.

File: applications/slonigiraf.md (hunk 4 above)
Comment (slonigiraf) [reply applications/slonigiraf.md:65]: The decline of memory retention in time is a well known phenomenon from psychology and pedagogy, see https://en.wikipedia.org/wiki/Forgetting_curve
Children loose some skills from the previous lesson thus in following lesson their previous teachers should be fined.

**Comment by slonigiraf** [comment]
> My last review's comments were not addressed, barring wiping the issue of correct incentive design under the rug of "all involved parties trust one another" - if they do, there's little need for a blockchain.

Blockchains always do some specific task. In our case it provides an ability to track teachers balance, fine teachers and see the validity of certificates. For sure such design can be done as a centralized app, but we're sure that no single company should be able to manipulate a reputation of people, thus we chose the decentralized version.
The phrase "All involved parties trust one another" is not worse than "DOT is something valuable thus we can stake it". You can't create a blockchain without people that use it, thus thoughts and beliefs of people rule how blockchain is operated. In our case teacher and student believe one another and student and employer believe one another. The system is used only by such people - not any random pair of people.

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the quick and detailed reply. To be honest, I’m still not convinced about the model of slashing teachers. Also I see another issue with your identity system. If I understand it correctly, students can basically create identities of teachers and therefore can add skills/certificates themselves as long as they have enough money (fictional teachers). Furthermore, I think it’s quite bold to immediately assume that you need your own relay chain for this project. But keep in mind the application is still ready for review and others might have a different opinion about the project.

**Comment by Lederstrumpf** [comment]
Hi @slonigiraf, thanks for your additional replies.
I'm unfortunately still left very sceptical of the underlying use case. In particular, there's cognitive dissonance wrt. the high amount of trust required to have this model work, and the amount of resources called for (relay + 9 parachains) is untenable, given the scope of this use case.

> The phrase "All involved parties trust one another" is not worse than "DOT is something valuable thus we can stake it". 

Yes it is worse.

My recommendation is that we close this PR.

File: applications/slonigiraf.md (hunk 5 above)
Comment (Lederstrumpf) [reply applications/slonigiraf.md:67]: I think you're modelling the demand for reimbursements incorrectly.

**Review by Lederstrumpf — CHANGES_REQUESTED** [review]

**Comment by slonigiraf** [comment]
Dear colleagues, I would be glad if any person from Web3 who has a teaching experience could review this grant application. By the now I hear only "the model will now work because I think so" from the reviewers. Please take into account that we have 3 years of experience using this educational model in real schools with real students and there is no question that the model itself is working and useful. The grant is about transferring this model as is onto Substrate based chain. We can use any other more friendly blockchain community. By now I think that Polkadot/Substrate is most promising community but you have your right to convince my in opposite.

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for pinging me again. I also pinged Bill Laboon here. I just want to point out that as far as I know, you have not tested this specific “educational model”, including the on-chain penalization and slashing, which is so far our biggest concern. Feel free to share reports/data that show otherwise.

**Comment by laboon** [comment]
@slonigiraf I'm a former lecturer in Computer Science at the University of Pittsburgh. Do you have a paper or study outlining the results from using this model in class? I've gone over the application and from an engineering and game theoretical perspective, share many of the same concerns as my colleagues here in terms of the actual implementation. 

I think there is a potentially good underlying idea here, i.e., staking reputation in a decentralized way for teaching. But -

1. I don't see how the real-world identity mappings are generated/verified in a decentralized way. If identity is not decentralized, what is the need for a blockchain? If it is decentralized, this seems like something one would need to dig more into or use a third-party DID service.
2. I don't understand what roles the Ed/Tanya/Teri/etc. characters are supposed to play - it seems like they were teacher/student/employer, but then it seems like other students are also judging students on a per-class basis?
3. Requiring up to nine parachains' worth of throughput for teaching a single country of students is... problematic. I would reconsider how much actual blockchain interaction there would have to be, although I think this an engineering problem. Since teachers for a given class are trusted, for example, class interactions can be bundled up and the hash stored on-chain, or using some sort of Merkel tree lookup etc. But this again goes into the decentralized identity aspect.
4. Running a separate relay chain along with parachains like  the model you suggest is going to be difficult for variety of reasons and would require a lot of extra engineering work. I think that minimizing the amount of L1 interactions would pay off more than preparing for this huge scale.

**Comment by slonigiraf** [comment]
@laboon many thanks for reviewing! Here are replies:

> * I don't see how the real-world identity mappings are generated/verified in a decentralized way. If identity is not decentralized, what is the need for a blockchain? If it is decentralized, this seems like something one would need to dig more into or use a third-party DID service.

Real-world identity mappings are generated by following: a students knows a teacher - it's a real world interaction. Student shows his public key to a teacher, teacher shows a public key to a student. It is the mapping. When the teacher signs a certificate about a skill for a student, the teacher signs with his private key a right to a student to allow employer fine a teacher. What force a teacher to give SLON to employer? The answer is a student - if teacher moves out it's SLON balance from the account before fine applied - both employer and a student know that the teacher cheated. The employer doesn't know who was it exactly - but actually he doesn't care - he just doesn't give a job to a student. But student will go to the teacher and ask - why he cheated. And as we suppose that initially the student and the teacher know each other and respect each other - the cheating should be rare event.
So what about identity mappings between a student and an employer? The student and the employer know each other in real world. The student respect employer as the employer is a public company with well known real world reputation. When the student come to the employer the student shows a certificate about his skill. Employer can re-exam the skill thus employer doesn't care if a student passed the exam by himself or he got the certificate by cheating. The employer just cares that the student is able to grant the right to impose fines and that the student has the skill - the employer re-exam the skill by himself. The only difference of this process to the current HR best practices is that the employer can get a SLON from cases when students miss the required skills. The employer then can sell SLON on an open marker and get money for the time he spent on a bad candidate.
Thus the identity is decentralized and there's no need in third party DID service.

> * I don't understand what roles the Ed/Tanya/Teri/etc. characters are supposed to play - it seems like they were teacher/student/employer, but then it seems like other students are also judging students on a per-class basis?

On a per-class basis the roles are following:
"employer" (Ed) = an adult teacher who work at school and run the class
"teacher" (Tanya/Teri) = any pupil/student within the class
"student" (Sarah) = any pupil within the class
Within the class motivation works as following.
Sarah wants to get a grade from Ed. The grade is a valuable thing because Sarah can show grade to her parents and get a praise for it.
To get a grade from Ed Sarah needs to show a list of certificates issued by Tery or Tania etc. Ed does have a time to monitor that certificates are valid but doesn't have enough time to teach all students. Thus Tery and Tania work as teachers, Ed does quality control of teaching. Any student can be in the role of Sarah at one time and in the role of Tery at another time.

> * Requiring up to nine parachains' worth of throughput for teaching a single country of students is... problematic. I would reconsider how much actual blockchain interaction there would have to be, although I think this an engineering problem. Since teachers for a given class are trusted, for example, class interactions can be bundled up and the hash stored on-chain, or using some sort of Merkle tree lookup etc. But this again goes into the decentralized identity aspect.

We do understand that getting 9 Polkadot/Kusama parachains is impossible thus we have our own relay for this purpose:
https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Fwss1.slonigiraf.org#/explorer

> * Running a separate relay chain along with parachains like  the model you suggest is going to be difficult for variety of reasons and would require a lot of extra engineering work. I think that minimizing the amount of L1 interactions would pay off more than preparing for this huge scale.

We have already launched our relay at April 12, 2021. Thus there is no extra engineering work here, it was already done. Here is a link to our relay:
https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Fwss1.slonigiraf.org#/explorer

**Comment by alxs** [comment]
@slonigiraf thank you for your extensive answers. I'm assuming the lack of a reply from @laboon means he isn't convinced, but he or anyone else can still get back to you or reopen the PR if they are interested. For the moment, seeing as the application has been in the review state for over a month now and hasn't received approval from any member of the committee, I'll go ahead and close it. Thank you for your interest in the grants program, and best of luck with your project.

**Review by Noc2 — CHANGES_REQUESTED** [review]
@slonigiraf Thanks for pinging me here. But it doesn’t look like you updated the application since it was reopened. Let me know if I’m missing anything.

**Review by Noc2 — CHANGES_REQUESTED** [review]
Maybe I misunderstood something, but I was under the impression that you talked with @laboon and you would potentially update the application and the scope of it again. I still have the same concerns as before regarding the current status of the application. Also, I have concerns regarding immediately launching a separate relay chain, which will be extremely difficult to maintain.

**Comment by slonigiraf** [comment]
> Noc2

Thanks for prompt reply. I will update the text in 2 days. By the way we've already launched the separate relay and it's working for 10 month from last April.

**Comment by slonigiraf** [comment]
Dear colleagues, @laboon , @Lederstrumpf , @Noc2. I've completely rewrote the grant application. First version was about education system and that lead to discussion. This version is more general - I request a financial support to create a recommendation letter system on blockchain that mimics 100% the real world recommendation letter system. Probably it would be better to submit it as a separate application but the actual coding rust work is the same - the same functions and the same data structures. Please review again, I think this version should be less controversial.

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the update. I just want to mention here that we are generally interested in people exploring on-chain reputation systems. Regarding your substrate module, I would assume that you require more [extrinsics](https://wiki.polkadot.network/docs/glossary#extrinsic) than the once, which you currently mention as part of your application. For example, how do you register a recommendation letter  in the first place? Also is there an on-chain system behind all the ids or are these simply hashes? To be honest, currently the application also seems to be a little bit expensive to me, if the pallet only comes with the two functions you mentioned.

**Comment by slonigiraf** [comment]
> Thanks for the update. I just want to mention here that we are generally interested in people exploring on-chain reputation systems. Regarding your substrate module, I would assume that you require more [extrinsics](https://wiki.polkadot.network/docs/glossary#extrinsic) than the once, which you currently mention as part of your application. For example, how do you register a recommendation letter in the first place? Also is there an on-chain system behind all the ids or are these simply hashes? To be honest, currently the application also seems to be a little bit expensive to me, if the pallet only comes with the two functions you mentioned.

@Noc2 thank you very much for your prompt reply. I've updated the application to answer you questions.
> "how do you register a recommendation letter in the first place"
A letter of recommendation is NOT registered on the blockchain when issued. This design principle saves blockchain space and hides important personal information from intruders. A letter of recommendation is a sign by guarantee of a string consisting of the letter_id, the public key of the guarantee, the public key of the worker, and the amount of reputation staked on this letter. The worker can sign the employer's right to send a letter of recommendation to the blockchain to penalize the guarantee. Thus we need only 1 extrinsic.
> "Also is there an on-chain system behind all the ids or are these simply hashes?"
There are hashes that represent guarantee, worker and employer account ids, and some of them are hashes of information that parties sign.
> "To be honest, currently the application also seems to be a little bit expensive to me, if the pallet only comes with the two functions"
For sure pallet will come with lot of functions and data structures. I've added the description of them in this commit.

**Review by Noc2 — CHANGES_REQUESTED** [review]
Thanks for the update. Got it. But I personally think that it’s still a little expensive for only a single pallet and the given functionality. Could you maybe reduce the price and/or integrate some additional deliveries, like a simple UI or weights?

**Comment by laboon** [comment]
I'd be interested in seeing some sort of prototype/MVP for this, even though I think there are still a few unanswered technical questions in terms of throughput. I think as development progresses they can be solved/ameliorated/changed (I would recommend looking at state channels or some other L2 or off-chain way to batch transactions together so more than one parachain is not necessary.

But 10k for a single developer for a month's work does seem expensive to me. I'm approving from a technical side, but I think @Noc2 would have a better idea of the right amount for the grant.

**Review by laboon — APPROVED** [review]
see comments above

**Review by laboon — APPROVED** [review]

**Review by masstez — APPROVED** [review]
